addons/l10n_hu_edi/models/l10n_hu_edi_transaction.py

```diff
@@ -0,0 +1,531 @@
+
+            transaction._process_query_transaction_result(invoice_result)
+
+    def _process_query_transaction_result(self, invoice_result):
+        def get_errors_from_invoice_result(invoice_result):
+            return [
+                f'({message["validation_result_code"]}) {message["validation_error_code"]}: {message["message"]}'
+                for message in invoice_result.get('business_validation_messages', []) + invoice_result.get('technical_validation_messages', [])
+            ]
+
+        self.ensure_one()
+        if invoice_result['invoice_status'] in ['RECEIVED', 'PROCESSING', 'SAVED']:
+            # The invoice has not been processed yet, so stay in state='sent'.
+            if self.state != 'sent':
+                # This is triggered if we come from the send_timeout state
+                self.write({
+                    'state': 'sent',
+                    'messages': {
+                        'error_title': _('The invoice was sent to the NAV, waiting for reply.'),
+                        'errors': [],
+                    },
+                })
+
+        elif invoice_result['invoice_status'] == 'DONE':
+            if not invoice_result['business_validation_messages'] and not invoice_result['technical_validation_messages']:
+                self.write({
+                    'state': 'confirmed',
+                    'messages': {
+                        'error_title': _('The invoice was successfully accepted by the NAV.'),
+                        'errors': [],
+                    },
+                })
+            else:
+                self.write({
+                    'state': 'confirmed_warning',
+                    'messages': {
+                        'error_title': _('The invoice was accepted by the NAV, but warnings were reported. To reverse, create a credit note.'),
+                        'errors': get_errors_from_invoice_result(invoice_result),
+                    },
+                })
+
+        elif invoice_result['invoice_status'] == 'ABORTED':
+            self.write({
+                'state': 'rejected',
+                'messages': {
+                    'error_title': _('The invoice was rejected by the NAV.'),
+                    'errors': get_errors_from_invoice_result(invoice_result),
+                },
+            })
+
+        else:
+            self.write({
+                'state': 'query_error',
+                'messages': {
+                    'error_title': _('NAV returned a non-standard invoice status: %s', invoice_result['invoice_status']),
+                    'errors': [],
+                },
+            })
+
+    @check_state_machine(action='recover_timeout')
+    def recover_timeout(self):
+        """ Attempt to recover all transactions in `self` from an upload timeout """
+        # Only attempt to recover from a timeout for transactions more than 6 minutes old.
+        self = self.filtered(lambda t: t.send_time <= fields.Datetime.now() - timedelta(minutes=6))
```

Comment:
It is an interesting way to check if something had timed out, but would not it give an error when you try to resend and then we can still check through this?

Reply:
I had a hard think about this. 

Of course, if you try to send an invoice and it already exists, you will get an error, so you know the invoice already exists. 

However, this is not enough to retrieve the existing invoice's state. The only way this API enables you to retrieve an invoice's state is if you have its *transaction ID* (there's no way to get the state by querying by invoice number). 

And, if you don't have an invoice's transaction ID, the only thing you can do is request a list of transaction IDs that were issued in a given time interval. 

So, basically, the only way you can retrieve an invoice's state, if you don't have its transaction ID, is if you know the time when it was sent. 

In the general case where you try to send an invoice and you get the duplicated error, you don't necessarily have accurate information about when the original invoice was sent, so it's unlikely you'd be able to recover from that.